packages/medusa-payment-stripe/src/api/hooks/stripe.ts

```diff
@@ -0,0 +1,161 @@
+import { Request, Response } from "express"
+import {
+  buildHandleCartPaymentErrorMessage,
+  constructWebhook,
+  isPaymentCollection,
+} from "../utils/utils"
+import {
+  AbstractCartCompletionStrategy,
+  CartService,
+  IdempotencyKeyService,
+} from "@medusajs/medusa"
+import { MedusaError } from "medusa-core-utils"
+
+export default async (req: Request, res: Response) => {
+  let event
+  try {
+    event = constructWebhook({
+      signature: req.headers["stripe-signature"],
+      body: req.body,
+      container: req.scope,
+    })
+  } catch (err) {
+    res.status(400).send(`Webhook Error: ${err.message}`)
+    return
+  }
+
+  const paymentIntent = event.data.object
+  const cartId = paymentIntent.metadata.cart_id // Backward compatibility
+  const resourceId = paymentIntent.metadata.resource_id
+
+  if (isPaymentCollection(resourceId)) {
+    await handlePaymentCollection(event, req, res, resourceId, paymentIntent.id)
+  } else {
+    await handleCartPayments(event, req, res, cartId ?? resourceId)
+  }
+}
```

Comment:
thought(non-blocking): As per [Stripe's documentation](https://stripe.com/docs/webhooks), we should attempt to respond with 200 as soon as possible and before doing any heavy work. This is to prevent Stripe from marking the event as failed, which will happen after [20 seconds without a response](https://stackoverflow.com/questions/71191152/stripe-webhook-best-practice). If our logic then fails, we are responsible for retrying that logic - not Stripe. 

That said, 20 seconds is plenty of time to perform the work here and we haven't experienced issues in the past, but I wanted to highlight it anyway to get your thoughts around it.

Reply:
Yes I ve seen this information, but I thought that 20s was plenty of time and using the stripe retry strategy would make sense since we have now the idempotency key. wdyt?

Reply:
When Stripe marks an event as failed, their retry mechanism will not kick in, so we don't really benefit from the idempotency key in those scenarios.

But as said, I am fine leaving it out for now and wait for issues to arise.

Reply:
The retry mechanism kick in if there is an error code 3XX or 4XX so unless we reach the timed out it will be retry. If we want to handle the time out we might want to leverage the event bus to also benefit from its retry mechanism. And the web hook function can be called in case of fail. wdyt?